django_graph_api/graphql/schema.py

```diff
@@ -1,14 +1,76 @@
 from collections import OrderedDict
+import copy
+
+from django.core.exceptions import ImproperlyConfigured
+import six
 
 from django_graph_api.graphql import introspection
 from django_graph_api.graphql.types import (
     Object,
+    ObjectMetaclass,
     RelatedField,
     String,
 )
 
 
-class BaseQueryRoot(Object):
+class CombinedQueryRootMetaclass(ObjectMetaclass):
+    def __new__(mcs, name, bases, attrs):
+        cls = super(CombinedQueryRootMetaclass, mcs).__new__(mcs, name, bases, attrs)
+
+        # Disallow explicit declared fields on combined query roots. They
+        # must only "inherit" fields.
+        if cls._declared_fields:
+            raise ImproperlyConfigured('CombinedQueryRoot may not declare fields')
+
+        # Disallow duplicate fields on query roots.
+        query_root_classes = attrs.get('query_root_classes', [])
+        seen_fields = {}
+        for query_root_class in query_root_classes:
+            for field_name in query_root_class._declared_fields:
+                seen_fields.setdefault(field_name, []).append(query_root_class)
+
+        if any(len(classes) > 1 for classes in seen_fields.values()):
+            error_str = '\n'.join(
+                '{} field is defined on multiple classes: {}'.format(
+                    field_name,
+                    ', '.join(query_root_class.__name__ for query_root_class in classes)
+                )
+                for field_name, classes in seen_fields.items()
+                if len(classes) > 1
+            )
+            raise ImproperlyConfigured(error_str)
+
+        # Collect declared fields from all query root classes.
+        for query_root_class in query_root_classes:
+            cls._declared_fields.update(copy.deepcopy(query_root_class._declared_fields))
+
+        return cls
+
+
+class CombinedQueryRoot(six.with_metaclass(CombinedQueryRootMetaclass, Object)):
+    def __init__(self, *args, **kwargs):
+        super(CombinedQueryRoot, self).__init__(*args, **kwargs)
+
+        self.query_roots = [
+            cls(*args, **kwargs)
+            for cls in (self.query_root_classes or [])
+        ]
+
+    def __getattr__(self, attr):
+        # Delegate resolvers to the class that defined the field if they aren't
+        # overridden on this class.
+        if attr[:4] == 'get_':
+            field_name = attr[4:]
+            for query_root in self.query_roots:
+                if field_name in query_root._declared_fields:
+                    return getattr(query_root, attr)
+        raise AttributeError('{} not found on {} or its query_root_classes'.format(
```

Comment:
I can't seem to get this error thrown.

If I define a field on a QueryRoot and don't define a resolver on it, line 66 runs and `getattr` throws an AttributeError, but in the Object's `get_raw_value` method, it just skips over it if it doesn't exist and then tries to return the model attribute, fails, and returns None.

I'm also not sure if we should be validating when you initialize the schema, because right now it will only run when you try to execute the field.

Reply:
This case is reached if the resolver call is for a field that doesn't exist (or an attribute that doesn't exist.) I don't think it's possible to get here with normal operation of the library but it could happen during debugging or if someone is using the library in an odd way.

I think this is the correct behavior but it should probably have a test and be written slightly differently...